src/octoprint/settings.py

```diff
@@ -543,17 +543,14 @@ def __str__(self):
 
 
 class HierarchicalChainMap(ChainMap):
-    def deep_dict(self, root=None):
-        if root is None:
-            root = self
 
-        result = {}
-        for key, value in root.items():
-            if isinstance(value, dict):
-                result[key] = self.deep_dict(root=self.__class__._get_next(key, root))
-            else:
-                result[key] = value
-        return result
+    def deep_dict(self):
```

Comment:
I looked into caching the results of this calculation, but I seemed like the cache was not being used. I'm not sure if it was being invalidated, or if we construct a bunch of these from scratch.

Reply:
I stored the cached value as `self._deep_dict_cache`, and cleared it whenever a write operation was performed. I saw no significant performance improvement.

Reply:
Can you take a look at eb69bd6fa90a01258063f3049f98a0335cbf7dbe that I just pushed and give it a try? I think that should achieve some caching, at least I saw several hits during page loads.

Reply:
That doesn't invalidate the cache often enough, sometimes `Settings` writes to `ChainMap.maps`. I implemented the caching in https://github.com/flaviut/OctoPrint/commit/788f6f97143ff3207663bda565a8713c2eb7b04d, and didn't see an improvement.

I also re-wrote the settings storage to essentially flatten the dictionaries as

```
{ a: b, c: {d: 1, e: 2}, f: {} }
a = b
c.d = 1
c.e = 2
f = {}
```

in https://github.com/flaviut/OctoPrint/commit/34004a299e8d59ba593363618e2dc7c754edd21e, but I haven't gotten a chance to test the performance of that, since I haven't been able to make it pass all the tests

Reply:
Oh right, I totally forgot about the overlays. I rolled the caching back for now. Let's concentrate on the failing tests.